src/distributions/range.rs

```diff
@@ -47,126 +45,451 @@ use distributions::Distribution;
+
+                (high, low)
+            }
+        }
+    }
+}
+
+#[cfg(not(feature = "i128_support"))]
+wmul_impl_large! { u64, 32 }
 #[cfg(feature = "i128_support")]
-integer_impl! { u128, u128 }
-integer_impl! { usize, usize }
+wmul_impl_large! { u128, 64 }
 
-macro_rules! float_impl {
+
+macro_rules! wmul_impl_usize {
     ($ty:ty) => {
+        impl WideningMultiply for usize {
+            type Output = (usize, usize);
+
+            #[inline(always)]
+            fn wmul(self, x: usize) -> Self::Output {
+                let (high, low) = (self as $ty).wmul(x as $ty);
+                (high as usize, low as usize)
+            }
+        }
+    }
+}
+
+#[cfg(target_pointer_width = "32")]
+wmul_impl_usize! { u32 }
+#[cfg(target_pointer_width = "64")]
+wmul_impl_usize! { u64 }
+
+
+
+/// Implementation of `RangeImpl` for float types.
+#[derive(Clone, Copy, Debug)]
+pub struct RangeFloat<X> {
+    scale: X,
+    offset: X,
+}
+
+macro_rules! range_float_impl {
+    ($ty:ty, $bits_to_discard:expr, $next_u:ident) => {
         impl SampleRange for $ty {
-            fn construct_range(low: $ty, high: $ty) -> Range<$ty> {
-                Range {
-                    low: low,
-                    range: high - low,
-                    accept_zone: 0.0 // unused
+            type T = RangeFloat<$ty>;
+        }
+
+        impl RangeImpl for RangeFloat<$ty> {
+            type X = $ty;
+
+            fn new(low: Self::X, high: Self::X) -> Self {
+                let scale = high - low;
+                let offset = low - scale;
+                RangeFloat {
+                    scale: scale,
+                    offset: offset,
                 }
             }
-            fn sample_range<R: Rng + ?Sized>(r: &Range<$ty>, rng: &mut R) -> $ty {
-                r.low + r.range * rng.gen::<$ty>()
+
+            fn new_inclusive(low: Self::X, high: Self::X) -> Self {
+                // Same as `new`, because the boundaries of a floats range are
+                // (at least currently) not exact due to rounding errors.
+                RangeImpl::new(low, high)
+            }
+
+            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
+                // Generate a value in the range [1, 2)
+                let value1_2 = (rng.$next_u() >> $bits_to_discard)
+                               .into_float_with_exponent(0);
+                // Doing multiply before addition allows some architectures to
+                // use a single instruction.
+                value1_2 * self.scale + self.offset
             }
         }
     }
 }
 
-float_impl! { f32 }
-float_impl! { f64 }
+range_float_impl! { f32, 32 - 23, next_u32 }
+range_float_impl! { f64, 64 - 52, next_u64 }
+
 
 #[cfg(test)]
 mod tests {
-    use distributions::Distribution;
-    use super::Range as Range;
+    use Rng;
+    use distributions::{Distribution};
+    use distributions::range::{Range, RangeImpl, RangeFloat, SampleRange};
+
+    #[test]
+    fn test_fn_range() {
```

Comment:
This test is pretty strange: why two separate loops? why not cache the ranges? do we also test the single-sample variant, and for various types?

Reply:
Oops, I copied the tests but didn't look closely. You are right, the first three test do not make much sense or are duplicates.